README
Number-Range-Regex version 0.31

This is a module which can be used to perform set operations on
ranges of integers efficiently, iterate over them, and convert them
to regular expressions. More information is available in the perldoc:
  perldoc lib/Number/Range/Regex.pm


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install


DEPENDENCIES

None.


COPYRIGHT AND LICENCE

Copyright (C) 2012, Brian Szymanski

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Perl extension Number::Range::Regexp

0.31 Oct 21 14:34:26 EDT 2012 r3171
   - various fixes and tests for negative numbers
   - init_opts defaults to default_opts, not empty hashref
   - save opts given at new() time in $self
   - add tests for invert(empty) == universal and invert(universal) == empty
   - fix bug in invert() on ranges not bounded with a max
   - multi_union() now returns an empty_range when given no arguments
   - fix some hard to track down edge cases in sectionify() that were
     triggered e.g. when doing set ops on ranges that touched but did
     not overlap
   - handle buggy rangespecs appropriately (eg "-3 -2..-1")
   - start using inf.pm to fix bugs with non-libc glibc (win32 and solaris
     at least):
       http://www.cpantesters.org/cpan/report/5e73b674-6ca0-1014-9aa7-d95518fa49ba
       http://www.cpantesters.org/cpan/report/f432d580-1921-11e2-bdcc-373e3b6b8117

0.30 Oct 17 11:41:24 EDT 2012 r3143
   - support for rangespec() with arguments of '-inf', '+inf'
   - implement & document is_infinite()
   - improve scope mangling in Util::most()
   - proper operations of all utility functions with infinite values
   - fix bug in xor affecting simple ranges
   - implement NRR::Util::inf to support arithmetic on potentially
     infinite values (perl's implementation is dodgy on 5.[68].X
     and with certain libc implementations)
   - restructure to eliminate a separate InfiniteRange.pm, saving
     ~150 lines of code
   - POD improvements

0.20 Oct  3 17:34:01 EDT 2012 r3074
   - fairly complete implementation - just missing intersect/subtract/xor
     on compound, infinite ranges.
   - support for negative integers
   - proper support for infinite ranges in all codepaths
   - new set operation: not()/invert()
   - allow whitespace in rangespecs: "3..5,7..9" == "3..5, 7..9"
   - fix regex_range() with options before call to init()
   - document NRR::Iterator, add $iterator->seek( number )
   - various under the hood improvements wrt maintenance

0.13  Sep 22 22:21:07 EDT 2012 r2970
   - graceful but ugly degradation: string context will give regex()
     output when overload can't distinguish between regex and string
     context (overload < v1.10, or equivalently, perl < v5.12)
   - more readable, less verbose CompoundRange regex commenting
       (since we include the CompoundRange::to_string() output,
       don't need it from each SimpleRange therein)
   - overhaul POD, documenting all functionality
   - support perl 5.6.x
   - overload.pm now required since it's been available with perl
     since 5.5.x or earlier)
   - regex_range() deprecated
   - add META.yml, README
   - less verbose regex commenting for CompoundRanges

0.12  Sep 21 00:36:38 EDT 2012 r2945
   -  overload ranges such that ->regex() and ->to_string() are implied
      depending on context (in a regex or string respectively), update
      POD and tests accordingly

0.11  Sep 20 23:37:45 EDT 2012 r2940
   -  regex_range( ... ) deprecated in favor of range( ... )->regex
   -  range & rangespec exported by default
   -  implement contains() to check for range membership w/o regexification
   -  document iterators
   -  common, fast implementation of all set ops (except union)
   -  implement to_string() to describe ranges
   -  EmptyRange->to_string() is now an empty string for consistency
   -  seemlessly promote SimpleRanges to CompoundRanges when necessary
   -  EmptyRange->regex() is now a regex that does not match anything
   -  unify option mangling code

0.09  Sep 19 18:31:36 EDT 2012 r2915
   -  NRR implements range() and rangespec() as recommended API
   -  intersection, subtraction, xor implemented
   -  EmptyRange->new()->regex() no longer dies - it now returns a regex
      that doesn't match anything
   -  clearer inheritance
   -  operate internally on SimpleRanges, not TrivialRanges. this allows
      for simpler code and faster operation (O(1) vs. O(lg n))
   -  implement iterators

0.08  Sep 10 15:47:15 EDT 2012 r2874
   -  object-orient-ify and implement union (intersect on the way)
   -  use string ops to avoid fp-induced bugs with large numbers
   -  thoroughly test 0..12 x 0..12 with regex_range()
   -  add some more targetted test cases
   -  refactor the two loops in regex_range to have parallel construction

0.07  Aug 29 21:23:23 2012 svn r2817
   -  readability: change e.g. '\d\d\d\d\d\d' to '\d{6}'
   -  $opts->{readable} implemented
   -  comment now works properly when min/max is undef

0.06  Aug 29 19:44:29 2012
   -  fix crash when regex_range() is called before init()

0.05  Tue Aug 28 22:20:42 2012
   -  silence a debug warning that was left enabled by accident

0.04  Tue Aug 28 21:44:08 2012 svn r2800
   -  auto-commenting implemented and enabled by default

0.03  Tue Aug 28 04:21:52 2012 svn r2794
   -  more exhaustive and informative test scripts

0.02  Thu Aug 23 13:45:23 2012
   -  no_leading_zeroes option implemented

0.01  Thu Aug 23 11:27:39 2012
   -  original version



lib/Number/Range/Regex.pm
# Number::Range::Regex

#
# Copyright 2012 Brian Szymanski.  All rights reserved.  This module is
# free software; you can redistribute it and/or modify it under the same
# terms as Perl itself.

package Number::Range::Regex;

use strict;
use Number::Range::Regex::Range;
use Number::Range::Regex::Iterator;
use Number::Range::Regex::Util;
use Number::Range::Regex::Util::inf qw( neg_inf pos_inf );
use vars qw ( @ISA @EXPORT @EXPORT_OK %EXPORT_TAGS $VERSION ); 
eval { require warnings; }; #it's ok if we can't load warnings

require Exporter;
use base 'Exporter';
@ISA    = qw( Exporter );
@EXPORT = qw( range rangespec );
@EXPORT_OK = qw ( init regex_range ) ;
%EXPORT_TAGS = ( all => [ @EXPORT, @EXPORT_OK ] );

$VERSION = '0.31';

my $init_opts = $Number::Range::Regex::Range::default_opts;

sub features { return { negative => 1 }; }

sub init {
  my ($self, @opts) = @_;

  # vestigial limb: init( foo => "bar" ) == init( { foo => "bar" } );
  my %opts = (@opts == 1) ? %{$opts[0]} :
             (@opts % 2 == 0) ? @opts :
             die 'usage: init( $options_ref )';

  $init_opts = $Number::Range::Regex::Range::default_opts;
  # override any values of init_opts that were passed to init
  while (my ($key, $value) = each %opts) {
    $init_opts->{$key} = $value;
  }
}

# regex_range( $min, $max ); #undef = no limit, so. e.g.
#   regex_range(3, undef) yields the equivalent of qr/[+]?[3-9]|\d+/;
sub regex_range {
  my ($min, $max, $passed_opts) = @_;

  # TODO: make options_mangler a little smarter for this case so it takes a
  #       default hashref arg.

  my $opts;
  if($passed_opts) {
    die "regex_range: too many arguments" unless ref $passed_opts eq 'HASH';
    $opts = { %$init_opts };
    while (my ($key, $val) = each %$passed_opts) {
      $opts->{$key} = $val;
    }
  } else {
    $opts = $init_opts;
  }

  return range($min, $max, $opts)->regex();
}

sub range {
  my ($min, $max, $opts) = @_;
  my $range;
  $min = neg_inf  if  !defined $min;
  $max = pos_inf  if  !defined $max;
  if($min == neg_inf && $max == pos_inf) {
    die "must specify either a min or a max or use the allow_wildcard argument" if !$opts->{allow_wildcard};
  }
  return Number::Range::Regex::SimpleRange->new( $min, $max, $opts );
}

sub rangespec {
  my $opts;
  if(ref $_[-1]) {
    $opts = option_mangler( pop );
  }
  # we allow (but do not like) e.g. rangespec(5,7,10..18);
  # we don't like it because it can make us run out of memory for
  # large ranges. preferred: rangespec('5,7,10..18');
  my $range;
  if(@_ > 1) {
    warn "passed literal range to rangespec!\n";
    $range = join ',', @_;
  } else {
    $range = $_[0];
  }

  my $section_validate  = qr/(?:-?\d+|(?:-?\d+|-inf)\.\.(?:\+?inf|-?\d+))/;
  my $range_validate = qr/(?:|$section_validate(?:,\s*$section_validate)*)/;
  die "invalid rangespec '$range'"  unless  $range =~ /^$range_validate$/;

  my @sections = split /,\s*/, $range;
  my @ranges;
  foreach my $section (@sections) {
    if($section =~ /^(-?\d+)$/) {
      push @ranges, Number::Range::Regex::SimpleRange->new( $1, $1, $opts );
    } else {
      my ($min, $max) = split /\.\./, $section, 2;
      if($min == neg_inf && $max == pos_inf && !$opts->{allow_wildcard}) {
        die "must specify either a min or a max or use the allow_wildcard argument";
      }
      $min = neg_inf  if  $min eq '-inf';
      $max = pos_inf  if  $max eq '+inf';
      push @ranges, Number::Range::Regex::SimpleRange->new( $min, $max, $opts );
    }
  }
  # note: multi_union() will have the side effect of sorting
  #       and de-overlap-ify-ing the input ranges
  return Number::Range::Regex::EmptyRange->new( $opts )  if  !@ranges;
  return multi_union( @ranges );
}

1;

__END__

=head1 NAME

Number::Range::Regex - create regular expressions that check for
                       integers in a given range

=head1 SYNOPSIS

  use Number::Range::Regex;
  my $lt_20 = range( 0, 19 );

  print "foo($foo) contains an integer < 20" if $foo =~ /$lt_20/;
  print "foo($foo) is an integer < 20" if $foo =~ /^$lt_20$/;
  if( $line =~ /^\S+\s+($lt_20)\s/ ) {
    print "the second field ($1) is an integer < 20";
  }
  my $nice_numbers = rangespec( "42,175..192" );
  my $my_values = $lt_20->union( $nice_numbers );
  if( $line =~ /^\S+\s+($my_values)\s/ ) {
    print "the second field has one of my values ($1)";
  }

  my $lt_10        = rangespec( "0..9" );
  my $primes_lt_30 = rangespec( "2,3,5,7,11,13,17,19,23,29" );
  my $primes_lt_10 = $lt_10->intersection( $primes_lt_30 );
  my $nonprimes_lt_10 = $lt_10->minus( $primes_lt_30 );
  print "nonprimes under 10 contains: ".join",", $nonprimes_lt_10->to_string;
  if( $something =~ /^$nonprimes_lt_10$/ ) {
    print "something($something) is a nonprime less than 10";
  }
  if( $nonprimes_lt_10->contains( $something ) ) {
    print "something($something) is a nonprime less than 10";
  }
  
  my $octet = range(0, 255);
  my $ip4_match = qr/^$octet\.$octet\.$octet\.$octet$/;
  my $range_96_to_127 = range(96, 127);
  my $my_slash26_match = qr/^192\.168\.42\.$range_96_to_127$/;
  my $my_slash19_match = qr/^192\.168\.$range_96_to_127\.$octet$/;

  my $in_a_or_in_b_but_not_both = $a->xor($b);

  my $it = rangespec("-20..42,47..52")->iterator();
  $it->first;
  do { print $it->fetch } while ($it->next);
  $it->last;
  do { print $it->fetch } while ($it->prev);
  

=head1 DESCRIPTION

Number::Range::Regex lets you manage sets of integers and generate
regular expressions matching them. For example, here is one way
to match number ranges in a regular expression:

  $date =~ m/^0*(?:[1-9]|[12][0-9]|3[01])\/0*(?:[0-9]|1[012])$/;

here is another:

  my $day_range = range(1, 31);
  my $month_range = range(1, 12);
  $date =~ m/^$day_range\/$month_range$/;

which is more legible? (bonus points if you spotted the bug)


=head1 METHODS

=head2 RANGE CREATION

=over

=item range

  $range = range( $min, $max );

Create a range between the first argument and the last. For example,
$min==8 and $max==12, corresponds to the list containing 8, 9, 10, 11,
and 12. This method is exported by default.

=item rangespec

  $range = rangespec( '8..12,14,19..22' );

Create a "compound" range given the range specification passed. For
example, the range above would consist of 8, 9, 10, 11, 12, 14, 19,
20, 21, and 22. This method is exported by default.

=back

=head2 RANGE INTERROGATION

=over

=item contains

  $range->contains( $number );

Returns a true value if $range contains $number - otherwise, it returns
a false value.

=item overlaps

  $range->overlaps( $another_range );

Returns a true value if $range overlaps with $another_range - otherwise,
it returns a false value. e.g.

  rangespec('7..9')->overlaps( rangespec('4..6') )   => false
  rangespec('7..9')->overlaps( rangespec('10..12') ) => false
  rangespec('7..9')->overlaps( rangespec('6..10') )  => true
  rangespec('7..9')->overlaps( rangespec('6..7') )   => true
  

=back

=head2 RANGE DISPLAY

=over

=item to_string

  $range->to_sting();

Return a compact representation of the range suitable for consumption
by a human, perl(1), or rangespec(). For example:

  $range = range( 6, 22 );
  print $range->to_string;

will output: "6..22", which can be parsed by perl(1) or rangespec().

=item regex

  $range->regex();

Return a regular expression matching members of this range. For example:

  $range = range( 6, 22 );
  print $range->regex;

will output something equivalent to:

  qr/0*(?:[6-9]|1\d|2[0-2])/

which, on my machine with perl v5.14.2 and a development version of
Number::Range::Regex between v0.12 and v0.13, is:

  (?^:(?# begin Number::Range::Regex::SimpleRange[6..22] )[+]?0*(?:(?^:[6-9])|(?^:1\d)|(?^:2[0-2]))(?# end Number::Range::Regex::SimpleRange[6..22] ))

=back

=head2 OVERLOADING

Please note that range objects are overloaded so that in regex
context, $range will be equivalent to $range->regex(). This
works in all versions of perl >= v5.6.0. When it is further
possible to distinguish regex context from string context (as
in overload v1.10 or higher, available in perl >= v5.12.0),
range objects will display in string but not regex context as
the terser, more legible $range->to_string() instead. 

If you find any cryptic errors about overloading, please use
an explicit ->to_string or ->regex() and file a bug.

=head2 UNBOUNDED (aka "infinite") RANGES

It is also possible to specify unbounded ranges, ie the set of
all integers less than 17. This may be specified in any of the
following ways:

  range( undef, 17 );
  range( '-inf', 17);
  rangespec('-inf..17');

Similarly the set of all integers greater than 17:

  range( 17, undef );
  range( 17, '+inf' );
  rangespec('17..+inf');

It is not, however, possible to specify the set of all integers by default.
If you try to do so, Number::Range::Regex will complain that you
"must specify either a min or a max or use the allow_wildcard argument":

  rangespec('-inf..+inf') #boom
  range( undef, '+inf' )  #boom
  range( undef, undef )   #boom
  range( '-inf', undef )  #let's go back to my room

however, the error you receive suggests the solution if you really
want such a range:

  range( '-inf', '+inf', {allow_wildcard => 1} ) 

To test if a range is infinite, you can call is_infinite():

  1 == ! rangespec('3..7')->is_infinite();
  1 == rangespec('16..+inf')->is_infinite();
  1 == rangespec('')->not->is_infinite();

=head2 SET OPERATIONS

given $range2 = rangespec( '0,2,4,6,8' ) and
      $range3 = rangespec( '0,3,6,9' )

=over

=item union

  $range = $range2->union( $range3 );

Return the union of one range with another. In the example above,
$range would consist of: 0, 2, 3, 4, 6, 8, and 9. Note that union()
can take more than one argument, e.g.
  $range2->union( $range3, $range5, $range7, $range11, ... );

=item intersect

  $range = $range2->intersect( $range3 );

Return the intersection of one range with another. In the example
above, $range would consist of: 0 and 6. This method is also available
via the alias intersection.

=item xor

  $range = $range->xor( $another_range );

Return the symmetric difference of $range2 and $range3 (elements in one
or the other range, but not in both). In the example above, $range would
consist of 2, 3, 4, 8, and 9.

=item subtract

  $range = $range2->subtract( $range3 );

Return the relative complement of $range2 in $range3. In the example
above, $range would consist of: 2, 4, and 8. Note carefully that
this method is not symmetric - $range3->subtract( $range2 ) would be
a different range consisting of 3 and 9. This method is also
available via the aliases subtraction and minus.

=item invert

  $range = $range2->invert();

Return the absolute complement of $range2. In the example above,
$range would include:

=over

  any integer less than or equal to -1
  1, 3, 5, and 7,  and
  any integer greater than or equal to 9.

=back

that is, $range->to_string would be '-inf..-1,1,3,5,7,9..+inf'. This
method is also available via the alias not.

=back

=head2 ITERATORS

iterators let you examine large or infinite sets with minimal memory:

  $it = $range->iterator();
  $it->first();
  do {
    do_something_with_value( $it->fetch );
  } while ($it->next);

for more information, see Number::Range::Regex::Iterator

=head2 OTHER METHODS

=item regex_range

=over

  $regex = regex_range( $min, $max );

This is a shortcut for range( $min, $max )->regex(). Useful for
one-off use when overload.pm does not support regex context. It
should only be used with perl 5.10.X or lower where regex context
overloading is not possible. This method may be deprecated in a
future release of Number::Range::Regex. This method is not exported
by default.

=back

=head1 NOTES

It's usually better to check for number-ness only in the regular
expression and verify the range of the number separately, eg:
  $line =~ /^\S+\s+(\d+)/ && $1 > 15 && $1 < 32;
but it's not always practical to refactor in that way.

If you like one-liners, something like the following may suit you...
  m{^${\( range(1, 31) )}\/${\( range(1, 12) )}$}
but, for readability's sake, please don't do that!


=head1 BUGS AND LIMITATIONS

Please report any bugs or feature requests through the
web interface at L<http://rt.cpan.org>.


=head1 AUTHOR

Brian Szymanski  B<< <ski-cpan@allafrica.com> >> -- be sure to put
Number::Range::Regex in the subject line if you want me to read
your message.


=head1 LICENSE

Copyright 2012 Brian Szymanski.  All rights reserved.  This module is
free software; you can redistribute it and/or modify it under the same
terms as Perl itself.


=head1 SEE ALSO

perl(1), Number::Range, etc.


=cut
